common/rfb/CSecurityARD.cxx

```diff
@@ -0,0 +1,150 @@
+/* 
+ * Copyright (C) 2022 Dinglan Peng
+ *    
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#ifndef HAVE_NETTLE
+#error "This header should not be compiled without HAVE_NETTLE defined"
+#endif
+
+#include <stdlib.h>
+#ifndef WIN32
+#include <unistd.h>
+#endif
+#include <assert.h>
+
+#include <nettle/aes.h>
+#include <nettle/md5.h>
+#include <nettle/bignum.h>
+#include <rfb/CSecurityARD.h>
+#include <rfb/CConnection.h>
+#include <rdr/InStream.h>
+#include <rdr/OutStream.h>
+#include <rfb/LogWriter.h>
+#include <rfb/Exception.h>
+#include <rfb/UserMsgBox.h>
+#include <os/os.h>
+
+using namespace rfb;
+
+const int MinKeyLength = 128;
+const int MaxKeyLength = 1024;
+
+CSecurityARD::CSecurityARD(CConnection* cc)
+  : CSecurity(cc), keyLength(0)
+{
+  mpz_init(g);
+  mpz_init(p);
+  mpz_init(A);
+  mpz_init(b);
+  mpz_init(B);
+  mpz_init(k);
+}
+
+CSecurityARD::~CSecurityARD()
+{
+  mpz_clear(g);
+  mpz_clear(p);
+  mpz_clear(A);
+  mpz_clear(b);
+  mpz_clear(B);
+  mpz_clear(k);
+}
+
+bool CSecurityARD::processMsg()
+{
+  if (readKey()) {
+    writeCredentials();
+    return true;
+  }
+  return false;
+}
+
+bool CSecurityARD::readKey()
+{
+  rdr::InStream* is = cc->getInStream();
+  if (!is->hasData(4))
+    return false;
+  is->setRestorePoint();
+  rdr::U16 gen = is->readU16();
+  keyLength = is->readU16();
+  if (keyLength < MinKeyLength)
+    throw AuthFailureException("DH key is too short");
+  if (keyLength > MaxKeyLength)
+    throw AuthFailureException("DH key is too long");

```

Comment:
Why the length checks? No such checks were added to the noVNC implementation of this.

Reply:
Sounds reasonable.